# Isoform analysis

Objective: compare the structure of isoforms within a gene using
grouping and disjoin operations.

There are many packages in Bioconductor that allow for isoform-level
analysis (also called transcript-level analysis). Packages that
facilitate differential transcript analysis include *DEXSeq* and
*DRIMSeq* (demonstrated in the *rnaseqDTU* workflow), and newer
packages *satuRn* and *fishpond*.

Once you've identified isoforms of interest within a gene, perhaps
isoforms that switch in terms of their expression after cells are
treated, one can use the
[IsoformSwitchAnalyzeR](https://github.com/kvittingseerup/IsoformSwitchAnalyzeR)
Bioconductor package to visualize and analyze a set of isoform
switches. For example, one can test the functional consequences of a
set of isoform switches in terms of the gain or loss of protein
domains, or splicing-centric changes (e.g. alternative 3' or 5'
acceptor sites, alternative transcription start or ends sites, etc.)

*IsoformSwitchAnalyzeR* is a multi-feature and mature package for this
type of analysis, but we can perform some simpler within-gene isoform
comparisons using *plyranges*, mostly for demonstration. We will start
again with the transcript database we've used before:

```{r message=FALSE}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
txp <- transcripts(txdb)
```

```{r message=FALSE}
library(plyranges)
txp <- txp %>%
  mutate(tx_id = as.character(tx_id)) %>%
  filter(seqnames == "chr1")
```

```{r}
txp <- txp %>%
  mutate(gene_id = mapIds(
           txdb, keys=tx_id,
           column="GENEID", keytype="TXID")
         ) %>%
  filter(!is.na(gene_id))
```

```{r}
library(org.Hs.eg.db)
txp <- txp %>%
  mutate(symbol = mapIds(
           org.Hs.eg.db, keys=gene_id,
           column="SYMBOL", keytype="ENTREZID")
         ) %>%
  filter(!is.na(symbol))
```

```{r message=FALSE}
library(tibble)
tab <- txp %>%
  group_by(gene_id) %>%
  summarize(ntxp = n_distinct(tx_id)) %>%
  as_tibble()
tab_longer <- dplyr::left_join(tibble(gene_id=txp$gene_id), tab)
txp <- txp %>%
  mutate(ntxp = tab_longer$ntxp)
```

```{r}
txp <- txp %>%
  filter(ntxp > 1)
```

```{r}
set.seed(1)
pick_one <- txp %>%
  as_tibble() %>%
  group_by(gene_id) %>%
  slice(sample.int(n(), size=1)) %>%
  dplyr::pull(tx_id)
```

```{r}
txp <- txp %>%
  mutate(the_one = as.integer(tx_id %in% pick_one))
```

```{r}
txp %>%
  group_by(gene_id) %>%
  disjoin_ranges(tx_ids = paste(tx_id,collapse=","))
```

```{r}
txp %>%
  group_by(gene_id) %>%
  disjoin_ranges(the_one_parts = min(the_one)) %>%
  filter(the_one_parts > 0)
```

```{r}
txp %>%
  filter(gene_id == "148413")
```

```{r}
ebt <- exonsBy(txdb, by="tx")
ebt <- ebt[txp$tx_id]
exons <- unlist(ebt) %>%
  select(exon_id, exon_rank) %>%
  mutate(tx_id = rep(names(ebt), lengths(ebt)))
exons
```

```{r}
txp_data <- txp %>%
  select(tx_id, gene_id, ntxp, the_one, .drop_ranges=TRUE) %>%
  as_tibble()
ids <- dplyr::left_join(tibble(tx_id = exons$tx_id),
                        txp_data, by="tx_id")
all.equal(exons$tx_id, ids$tx_id)
mcols(exons) <- cbind(mcols(exons), ids %>% select(-tx_id))
exons
```

```{r}
exon_parts <- exons %>%
  group_by(gene_id) %>%
  disjoin_ranges(the_one_parts = min(the_one)) %>%
  filter(the_one_parts > 0)
exon_parts
```

```{r}
txp %>%
  filter(gene_id == "339451")
tx_to_show <- txp %>%
  filter(gene_id == "339451" & the_one == 1) %>%
  as_tibble() %>%
  dplyr::pull(tx_name)
these_parts <- exon_parts %>%
  filter(gene_id == "339451")
```

```{r message=FALSE}
library(plotgardener)
par <- pgParams(
  chrom = "chr1", 
  chromstart = 895.9e3, chromend = 901.2e3,
  assembly = "hg19", just = c("left", "bottom")
)
```

```{r}
hilite <- data.frame(transcript=tx_to_show, color="magenta")
```

```{r isoform-plot}
pageCreate(width = 5, height = 2.5, showGuides = TRUE)
plotTranscripts(
  params = par, x = 0.5, y = 1.5, width = 4, height = 1.5,
  transcriptHighlights = hilite
)
plotRanges(
  these_parts, fill="darkorchid",
  params = par, x = 0.5, y = 1.75, width = 4, height = .25
)
plotText(
  label = paste("unique pieces of",tx_to_show), fontcolor = "darkorchid",
  params = par, x = 3.1, y = 1.75,
  just = c("left", "bottom"), fontsize = 8
)
plotGenomeLabel(
  params = par, x = 0.5, y = 2, length = 4,
  just = c("left", "top")
)
```

Questions:

* How else could we have found pieces of one isoform per gene, that do
  not belong to any other isoforms of the genes. Would other
  approaches have any limitations?
