# Tidy RNA-seq EDA

Objective: learn how to make basic EDA plots (plotting scaled counts)
of RNA-seq data using tidy-style verbs.

The *tidybulk* Bioconductor package and associated
[tidy transcriptomics](https://stemangiola.github.io/tidytranscriptomics/post/tidy-transcriptomics-manifesto/) 
ecosystem (tidySummarizedExperiment, tidySingleCell, tidyseurat, etc.)
provide a bridge between the *SummarizedExperiment*-style
representation of matrix data with attached metadata, to the
tidy-style representation of data (one row per observation).
For more details on the package, check out the 
[tidybulk](https://stemangiola.github.io/tidybulk/) 
website, and the publication: @tidybulk. 

Here we will briefly examine the difference between tidy manipulation
of these matrix-style objects compared to how these objects would be
manipulated in other Bioconductor workflows (e.g. DESeq2, edgeR, or
limma workflows). As you will see, there are multiple ways to generate
the same or similar exploratory plots and analyses.  we recommend to
consider which is more appropriate to your purpose. For package code,
you may prefer to have fewer dependencies, going with core
Bioconductor objects and packages. For scripting, the "fluent" and
piped style of *tidybulk* may be preferable, and easier for others to
read and modify your code at a future date.

Here we just compare code for some basic tasks, scaling counts and
performing DE with DESeq2. However, note that *tidybulk* has
many functionalities implemented, including dimension reduction and
visualization (PCA, MDS, tSNE, UMAP), clustering, gene set testing,
cell type composition analysis and cell abundance testing, unwanted
variation modeling, imputation, etc.

```{r}
library(oct4)
dir <- system.file("extdata", package="oct4")
coldata <- read.csv(file.path(dir,"coldata.csv"))
coldata
coldata$files <- file.path(dir, "quants", coldata$names, "quant.sf.gz")
```

```{r eval=FALSE}
library(tximeta)
se <- tximeta(coldata)
gse <- summarizeToGene(se)
library(org.Mm.eg.db)
gse <- addIds(gse, "SYMBOL")
```

```{r eval=FALSE}
library(SummarizedExperiment)
assayNames(gse)
assays(gse) <- assays(gse)[1:3]
# save for easy loading later
save(gse, file="data/oct4_obj.rda")
```

```{r message=FALSE}
library(SummarizedExperiment)
load("data/oct4_obj.rda")
```

```{r}
gse$rep <- rep(1:3, 4)
colnames(gse) <- paste(gse$line,gse$condition,gse$rep,sep="-")
assay(gse, "counts") <- round(assay(gse, "counts")) # for consistency
```

```{r}
library(AnnotationDbi)
library(org.Mm.eg.db)
# pluripotency
tab <- select(org.Mm.eg.db, "GO:0019827", "SYMBOL", "GO")
tab <- tab[!duplicated(tab$SYMBOL),]
pluri <- tab$SYMBOL
```

```{r message=FALSE}
library(tidybulk)
library(dplyr)
library(stringr)
oct4 <- gse %>%
  tidybulk() %>%
  filter(line == "OCT4") %>%
  mutate(.sample = str_remove(.sample, "OCT4-")) %>%
  mutate(.sample = factor(.sample, levels=unique(.sample)),
         condition = factor(condition, c("untrt","trt")))
```

```{r}
oct4 <- oct4 %>%
  keep_abundant(factor_of_interest = condition) %>%
  scale_abundance(method="RLE") # DESeq2 scaling
```

```{r tidy-boxplot}
library(ggplot2)
oct4 %>%
  ggplot(aes(.sample, log10(counts_scaled + 1))) +
  geom_boxplot()
```

```{r deseq-boxplot}
library(DESeq2)
gse_sub <- gse[ , gse$line == "OCT4" ]
gse_sub$condition <- factor(gse_sub$condition)
dds <- gse_sub %>%
  DESeqDataSet(~condition) %>%
  estimateSizeFactors()
boxplot(log10(counts(dds, normalized=TRUE) + 1))
```

```{r tidy-lines}
oct4 %>%
  filter(SYMBOL %in% pluri) %>%
  mutate(logcounts = log10(counts_scaled + 1)) %>%
  mutate(Oct4 = ifelse(SYMBOL == "Pou5f1", "red", "black")) %>%
  group_by(.feature) %>%
  mutate(logcounts = logcounts - mean(logcounts)) %>%
  ungroup() %>%
  ggplot(aes(.sample, logcounts, group=.feature, color=Oct4)) +
  geom_point() +
  geom_line() +
  scale_color_identity()
```

```{r deseq-lines}
gene_idx <- mcols(dds)$SYMBOL %in% pluri
mat <- log10(counts(dds, normalized=TRUE)[gene_idx,] + 1)
mat <- mat - rowMeans(mat)
hilite <- rownames(dds)[which(mcols(dds)$SYMBOL == "Pou5f1")]
plot(mat[1,], type="n", ylim=c(-1,1), xlab="samples", ylab="logcounts")
for (i in 1:nrow(mat)) {
  col <- ifelse(rownames(mat)[i] == hilite, "red", "black")
  points(mat[i,], type="b", col=col)
}
```

```{r}
gene_idx <- oct4 %>% pivot_transcript() %>% pull(.feature)
assays(dds) <- assays(dds)[1:2] # compatiblity with tidybulk
res <- dds[gene_idx,] %>%
  DESeq(quiet=TRUE) %>%
  results()
```

```{r}
oct4 <- oct4 %>%
  test_differential_abundance(~condition, method="deseq2", quiet=TRUE)
tidy_res <- oct4 %>%
  pivot_transcript()
```

```{r}
all.equal(rownames(res), tidy_res$.feature)
table(base_sig = res$padj < .1, tidy_sig = tidy_res$padj < .1)
```

```{r}
plot_data <- oct4 %>%
  filter(SYMBOL %in% pluri) %>%
  mutate(logcounts = log10(counts_scaled + 1)) %>%
  mutate(Oct4 = ifelse(SYMBOL == "Pou5f1", "red", "black")) %>%
  group_by(.feature) %>%
  mutate(logcounts = logcounts - mean(logcounts)) %>%
  ungroup() %>%
  mutate(gene_type = case_when(
           padj < .1 & log2FoldChange > 0 ~ "up",
           padj < .1 & log2FoldChange < 0 ~ "down",
           TRUE ~ "null"))
```

```{r tidy-lines-faceted}
library(ggrepel)
plot_data %>%
  filter(gene_type != "null") %>%
  ggplot(aes(.sample, logcounts, group=.feature, color=Oct4)) +
  geom_point() +
  geom_line() +
  geom_text_repel(data=plot_data %>%
                    filter(.sample == "trt-3", gene_type != "null"),
            aes(.sample, logcounts, label=SYMBOL), nudge_x=.5) +
  scale_color_identity() +
  facet_wrap(~gene_type) +
  scale_x_discrete(expand = expansion(add = 2)) +
  xlab("sample")
```
