<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Gene plots | Tidy Ranges Tutorial</title>
  <meta name="description" content="<p>Basic examples of computing operations on genomic ranges using the
tidy data philosophy.</p>" />
  <meta name="generator" content="bookdown 0.32 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Gene plots | Tidy Ranges Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="<p>Basic examples of computing operations on genomic ranges using the
tidy data philosophy.</p>" />
  <meta name="github-repo" content="nullranges/tidy-ranges-tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Gene plots | Tidy Ranges Tutorial" />
  
  <meta name="twitter:description" content="<p>Basic examples of computing operations on genomic ranges using the
tidy data philosophy.</p>" />
  

<meta name="author" content="nullranges devel team" />


<meta name="date" content="2023-05-09" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="snp-position-in-peaks.html"/>
<link rel="next" href="rna-seq-eda.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Tidy Ranges Tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a></li>
<li class="chapter" data-level="1" data-path="join-is-an-overlap.html"><a href="join-is-an-overlap.html"><i class="fa fa-check"></i><b>1</b> Join is an overlap</a></li>
<li class="chapter" data-level="2" data-path="compute-overlaps.html"><a href="compute-overlaps.html"><i class="fa fa-check"></i><b>2</b> Compute overlaps</a></li>
<li class="chapter" data-level="3" data-path="promoter-marks.html"><a href="promoter-marks.html"><i class="fa fa-check"></i><b>3</b> Promoter marks</a></li>
<li class="chapter" data-level="4" data-path="bootstrap-overlap.html"><a href="bootstrap-overlap.html"><i class="fa fa-check"></i><b>4</b> Bootstrap overlap</a></li>
<li class="chapter" data-level="5" data-path="snp-position-in-peaks.html"><a href="snp-position-in-peaks.html"><i class="fa fa-check"></i><b>5</b> SNP position in peaks</a></li>
<li class="chapter" data-level="6" data-path="gene-plots.html"><a href="gene-plots.html"><i class="fa fa-check"></i><b>6</b> Gene plots</a></li>
<li class="chapter" data-level="7" data-path="rna-seq-eda.html"><a href="rna-seq-eda.html"><i class="fa fa-check"></i><b>7</b> RNA-seq EDA</a></li>
<li class="chapter" data-level="8" data-path="many-models.html"><a href="many-models.html"><i class="fa fa-check"></i><b>8</b> Many models</a></li>
<li class="chapter" data-level="9" data-path="isoform-analysis.html"><a href="isoform-analysis.html"><i class="fa fa-check"></i><b>9</b> Isoform analysis</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tidy Ranges Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="gene-plots" class="section level1 hasAnchor" number="6">
<h1><span class="header-section-number">Chapter 6</span> Gene plots<a href="gene-plots.html#gene-plots" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Objective: for a region of the genome, find peaks near the TSS of genes
and then plot their signal strength per gene, stratifying by the tissue
origin of the peak.</p>
<p>We start by loading the pre-downloaded peaks ranges:</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="gene-plots.html#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;data/peaks.rda&quot;</span>)</span></code></pre></div>
<p>Likewise, we want to use hg19 genes again to match the hg19 peaks:</p>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="gene-plots.html#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)</span>
<span id="cb138-2"><a href="gene-plots.html#cb138-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb138-3"><a href="gene-plots.html#cb138-3" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">genes</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)</span></code></pre></div>
<p>Adding gene symbols:</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="gene-plots.html#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyranges)</span>
<span id="cb139-2"><a href="gene-plots.html#cb139-2" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> g <span class="sc">%&gt;%</span></span>
<span id="cb139-3"><a href="gene-plots.html#cb139-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene_name =</span> <span class="fu">mapIds</span>(org.Hs.eg.db, </span>
<span id="cb139-4"><a href="gene-plots.html#cb139-4" aria-hidden="true" tabindex="-1"></a>                            gene_id, <span class="st">&quot;SYMBOL&quot;</span>, <span class="st">&quot;ENTREZID&quot;</span>))</span></code></pre></div>
<p>Find a region of the genome near a kidney-specific gene:</p>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="gene-plots.html#cb140-1" aria-hidden="true" tabindex="-1"></a>g <span class="sc">%&gt;%</span> </span>
<span id="cb140-2"><a href="gene-plots.html#cb140-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene_name <span class="sc">==</span> <span class="st">&quot;UMOD&quot;</span>)</span></code></pre></div>
<pre><code>## GRanges object with 1 range and 2 metadata columns:
##        seqnames            ranges strand |     gene_id   gene_name
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;
##   7369    chr16 20344373-20364037      - |        7369        UMOD
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="gene-plots.html#cb142-1" aria-hidden="true" tabindex="-1"></a>region <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb142-2"><a href="gene-plots.html#cb142-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">seqnames=</span><span class="st">&quot;chr16&quot;</span>, </span>
<span id="cb142-3"><a href="gene-plots.html#cb142-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">start=</span><span class="fl">20e6</span>,</span>
<span id="cb142-4"><a href="gene-plots.html#cb142-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">end=</span><span class="fl">21e6</span>) <span class="sc">%&gt;%</span></span>
<span id="cb142-5"><a href="gene-plots.html#cb142-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_granges</span>()</span></code></pre></div>
<p>Combine the bladder and kidney peaks, and select certain columns:</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="gene-plots.html#cb143-1" aria-hidden="true" tabindex="-1"></a>pks <span class="ot">&lt;-</span> <span class="fu">bind_ranges</span>(<span class="at">bladder=</span>bladder_pks,</span>
<span id="cb143-2"><a href="gene-plots.html#cb143-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">kidney=</span>kidney_pks,</span>
<span id="cb143-3"><a href="gene-plots.html#cb143-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.id=</span><span class="st">&quot;tissue&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb143-4"><a href="gene-plots.html#cb143-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="at">signal=</span>signalValue, tissue)</span></code></pre></div>
<p>Finally, we perform the overlap join, locating peaks within 100kb of
the TSS of the gene.</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="gene-plots.html#cb144-1" aria-hidden="true" tabindex="-1"></a>g_with_pks <span class="ot">&lt;-</span> g <span class="sc">%&gt;%</span></span>
<span id="cb144-2"><a href="gene-plots.html#cb144-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">anchor_5p</span>() <span class="sc">%&gt;%</span></span>
<span id="cb144-3"><a href="gene-plots.html#cb144-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">width=</span><span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb144-4"><a href="gene-plots.html#cb144-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_by_overlaps</span>(region) <span class="sc">%&gt;%</span></span>
<span id="cb144-5"><a href="gene-plots.html#cb144-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">join_overlap_inner</span>(pks, <span class="at">maxgap=</span><span class="fl">1e5</span>)</span>
<span id="cb144-6"><a href="gene-plots.html#cb144-6" aria-hidden="true" tabindex="-1"></a>g_with_pks<span class="sc">$</span>tissue <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span></code></pre></div>
<pre><code>## .
## bladder  kidney 
##     208     337</code></pre>
<p>We can construct a faceted set of boxplots, first we make a tibble of
data for our plot.</p>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="gene-plots.html#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb146-2"><a href="gene-plots.html#cb146-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb146-3"><a href="gene-plots.html#cb146-3" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> g_with_pks <span class="sc">%&gt;%</span></span>
<span id="cb146-4"><a href="gene-plots.html#cb146-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(gene_name, signal, tissue, <span class="at">.drop_ranges=</span><span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb146-5"><a href="gene-plots.html#cb146-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>()</span></code></pre></div>
<p>Then pass the data to ggplot (we could have just passed the data
directly, but we plan to re-use the data).</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="gene-plots.html#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb147-2"><a href="gene-plots.html#cb147-2" aria-hidden="true" tabindex="-1"></a>dat <span class="sc">%&gt;%</span></span>
<span id="cb147-3"><a href="gene-plots.html#cb147-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(tissue, signal)) <span class="sc">+</span></span>
<span id="cb147-4"><a href="gene-plots.html#cb147-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span> </span>
<span id="cb147-5"><a href="gene-plots.html#cb147-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>gene_name)</span></code></pre></div>
<p><img src="gene-plots_files/figure-html/peaks-near-genes-1.png" width="672" /></p>
<p>Now let’s try to plot these in context, using <em>plotgardener</em>
<span class="citation">(<a href="#ref-plotgardener" role="doc-biblioref">Kramer et al. 2022</a>)</span>.
First we filter down to the peaks near <em>UMOD</em>.</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="gene-plots.html#cb148-1" aria-hidden="true" tabindex="-1"></a>umod <span class="ot">&lt;-</span> g <span class="sc">%&gt;%</span></span>
<span id="cb148-2"><a href="gene-plots.html#cb148-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene_name <span class="sc">==</span> <span class="st">&quot;UMOD&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb148-3"><a href="gene-plots.html#cb148-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">anchor_5p</span>() <span class="sc">%&gt;%</span></span>
<span id="cb148-4"><a href="gene-plots.html#cb148-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">width=</span><span class="dv">1</span>)</span>
<span id="cb148-5"><a href="gene-plots.html#cb148-5" aria-hidden="true" tabindex="-1"></a>pks_to_plot <span class="ot">&lt;-</span> pks <span class="sc">%&gt;%</span></span>
<span id="cb148-6"><a href="gene-plots.html#cb148-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter_by_overlaps</span>(umod, <span class="at">maxgap=</span><span class="fl">1e5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb148-7"><a href="gene-plots.html#cb148-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">anchor_center</span>() <span class="sc">%&gt;%</span></span>
<span id="cb148-8"><a href="gene-plots.html#cb148-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">width=</span><span class="fl">1e4</span>) <span class="co"># to make the ranges more visible</span></span></code></pre></div>
<p>We then define a color scheme for the tissue variable, and make a
ggplot object which will be added to our genome plots.</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="gene-plots.html#cb149-1" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="cf">function</span>(n) <span class="fu">palette.colors</span>(n<span class="sc">+</span><span class="dv">2</span>)[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>)]</span>
<span id="cb149-2"><a href="gene-plots.html#cb149-2" aria-hidden="true" tabindex="-1"></a>col_vec <span class="ot">&lt;-</span> <span class="fu">cols</span>(<span class="dv">2</span>)</span>
<span id="cb149-3"><a href="gene-plots.html#cb149-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(col_vec) <span class="ot">&lt;-</span> <span class="fu">unique</span>(dat<span class="sc">$</span>tissue)</span>
<span id="cb149-4"><a href="gene-plots.html#cb149-4" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> dat <span class="sc">%&gt;%</span></span>
<span id="cb149-5"><a href="gene-plots.html#cb149-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gene_name <span class="sc">==</span> <span class="st">&quot;UMOD&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb149-6"><a href="gene-plots.html#cb149-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(tissue, signal, <span class="at">col=</span>tissue)) <span class="sc">+</span></span>
<span id="cb149-7"><a href="gene-plots.html#cb149-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># here we set a seed for jitter</span></span>
<span id="cb149-8"><a href="gene-plots.html#cb149-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>, <span class="at">position=</span><span class="fu">position_jitter</span>(<span class="at">width=</span>.<span class="dv">1</span>, <span class="at">seed=</span><span class="dv">5</span>)) <span class="sc">+</span></span>
<span id="cb149-9"><a href="gene-plots.html#cb149-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> col_vec)</span></code></pre></div>
<p>We then create some parameters that will be shared across a number of
the plots in <em>plotgardener</em>.</p>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="gene-plots.html#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plotgardener)</span>
<span id="cb150-2"><a href="gene-plots.html#cb150-2" aria-hidden="true" tabindex="-1"></a>par <span class="ot">&lt;-</span> <span class="fu">pgParams</span>(</span>
<span id="cb150-3"><a href="gene-plots.html#cb150-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">chrom =</span> <span class="st">&quot;chr16&quot;</span>, </span>
<span id="cb150-4"><a href="gene-plots.html#cb150-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">chromstart =</span> <span class="fu">round</span>((<span class="fu">start</span>(umod) <span class="sc">-</span> <span class="fl">1e5</span>)<span class="sc">/</span><span class="fl">1e5</span>)<span class="sc">*</span><span class="fl">1e5</span>,</span>
<span id="cb150-5"><a href="gene-plots.html#cb150-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">chromend =</span> <span class="fu">round</span>((<span class="fu">start</span>(umod) <span class="sc">+</span> <span class="fl">4e5</span>)<span class="sc">/</span><span class="fl">1e5</span>)<span class="sc">*</span><span class="fl">1e5</span>,</span>
<span id="cb150-6"><a href="gene-plots.html#cb150-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">assembly =</span> <span class="st">&quot;hg19&quot;</span>, <span class="at">just =</span> <span class="fu">c</span>(<span class="st">&quot;left&quot;</span>, <span class="st">&quot;bottom&quot;</span>)</span>
<span id="cb150-7"><a href="gene-plots.html#cb150-7" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Finally we put all the pieces together on a page (for laying out the
plot, first use <code>showGuides=TRUE</code>).</p>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="gene-plots.html#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pageCreate</span>(<span class="at">width =</span> <span class="dv">5</span>, <span class="at">height =</span> <span class="dv">3</span>, <span class="at">showGuides =</span> <span class="cn">FALSE</span>)</span>
<span id="cb151-2"><a href="gene-plots.html#cb151-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenes</span>(</span>
<span id="cb151-3"><a href="gene-plots.html#cb151-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">params =</span> par, <span class="at">x =</span> <span class="fl">0.5</span>, <span class="at">y =</span> <span class="fl">2.5</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> .<span class="dv">75</span></span>
<span id="cb151-4"><a href="gene-plots.html#cb151-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb151-5"><a href="gene-plots.html#cb151-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRanges</span>(</span>
<span id="cb151-6"><a href="gene-plots.html#cb151-6" aria-hidden="true" tabindex="-1"></a>  pks_to_plot,</span>
<span id="cb151-7"><a href="gene-plots.html#cb151-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">colorby</span>(<span class="st">&quot;tissue&quot;</span>, <span class="at">palette=</span>cols),</span>
<span id="cb151-8"><a href="gene-plots.html#cb151-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">params =</span> par, <span class="at">x =</span> <span class="fl">0.5</span>, <span class="at">y =</span> <span class="fl">1.75</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">1.75</span></span>
<span id="cb151-9"><a href="gene-plots.html#cb151-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb151-10"><a href="gene-plots.html#cb151-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGenomeLabel</span>(</span>
<span id="cb151-11"><a href="gene-plots.html#cb151-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">params =</span> par, <span class="at">x =</span> <span class="fl">0.5</span>, <span class="at">y =</span> <span class="fl">2.5</span>, <span class="at">length =</span> <span class="dv">4</span>,</span>
<span id="cb151-12"><a href="gene-plots.html#cb151-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">just =</span> <span class="fu">c</span>(<span class="st">&quot;left&quot;</span>, <span class="st">&quot;top&quot;</span>)</span>
<span id="cb151-13"><a href="gene-plots.html#cb151-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb151-14"><a href="gene-plots.html#cb151-14" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGG</span>(</span>
<span id="cb151-15"><a href="gene-plots.html#cb151-15" aria-hidden="true" tabindex="-1"></a>  p <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;UMOD&quot;</span>), </span>
<span id="cb151-16"><a href="gene-plots.html#cb151-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">params =</span> par, <span class="at">x =</span> <span class="fl">2.25</span>, <span class="at">y =</span> <span class="fl">1.75</span>, <span class="at">width =</span> <span class="fl">2.5</span>, <span class="at">height =</span> <span class="fl">1.75</span></span>
<span id="cb151-17"><a href="gene-plots.html#cb151-17" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="gene-plots_files/figure-html/garden-1.png" width="480" /></p>

</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-plotgardener" class="csl-entry">
Kramer, Nicole E, Eric S Davis, Craig D Wenger, Erika M Deoudes, Sarah M Parker, Michael I Love, and Douglas H Phanstiel. 2022. <span>“<span class="nocase">Plotgardener: cultivating precise multi-panel figures in R</span>.”</span> <em>Bioinformatics</em> 38 (7): 2042–45. <a href="https://doi.org/10.1093/bioinformatics/btac057">https://doi.org/10.1093/bioinformatics/btac057</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="snp-position-in-peaks.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="rna-seq-eda.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
