<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Promoter marks | Tidy Ranges Tutorial</title>
  <meta name="description" content="<p>Basic examples of computing operations on genomic ranges using the
tidy data philosophy.</p>" />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Promoter marks | Tidy Ranges Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="<p>Basic examples of computing operations on genomic ranges using the
tidy data philosophy.</p>" />
  <meta name="github-repo" content="tidybiology/tidy-ranges-tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Promoter marks | Tidy Ranges Tutorial" />
  
  <meta name="twitter:description" content="<p>Basic examples of computing operations on genomic ranges using the
tidy data philosophy.</p>" />
  

<meta name="author" content="Michael Love" />


<meta name="date" content="2023-08-01" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="compute-overlaps.html"/>
<link rel="next" href="bootstrap-overlap.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Tidy Ranges Tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a></li>
<li class="chapter" data-level="1" data-path="join-is-an-overlap.html"><a href="join-is-an-overlap.html"><i class="fa fa-check"></i><b>1</b> Join is an overlap</a></li>
<li class="chapter" data-level="2" data-path="compute-overlaps.html"><a href="compute-overlaps.html"><i class="fa fa-check"></i><b>2</b> Compute overlaps</a></li>
<li class="chapter" data-level="3" data-path="promoter-marks.html"><a href="promoter-marks.html"><i class="fa fa-check"></i><b>3</b> Promoter marks</a></li>
<li class="chapter" data-level="4" data-path="bootstrap-overlap.html"><a href="bootstrap-overlap.html"><i class="fa fa-check"></i><b>4</b> Bootstrap overlap</a></li>
<li class="chapter" data-level="5" data-path="snp-position-in-peaks.html"><a href="snp-position-in-peaks.html"><i class="fa fa-check"></i><b>5</b> SNP position in peaks</a></li>
<li class="chapter" data-level="6" data-path="gene-plots.html"><a href="gene-plots.html"><i class="fa fa-check"></i><b>6</b> Gene plots</a></li>
<li class="chapter" data-level="7" data-path="rna-seq-eda.html"><a href="rna-seq-eda.html"><i class="fa fa-check"></i><b>7</b> RNA-seq EDA</a></li>
<li class="chapter" data-level="8" data-path="many-genomic-models.html"><a href="many-genomic-models.html"><i class="fa fa-check"></i><b>8</b> Many genomic models</a></li>
<li class="chapter" data-level="9" data-path="isoform-analysis.html"><a href="isoform-analysis.html"><i class="fa fa-check"></i><b>9</b> Isoform analysis</a></li>
<li class="chapter" data-level="10" data-path="gene-ranges-in-tximeta.html"><a href="gene-ranges-in-tximeta.html"><i class="fa fa-check"></i><b>10</b> Gene ranges in tximeta</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tidy Ranges Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="promoter-marks" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">Chapter 3</span> Promoter marks<a href="promoter-marks.html#promoter-marks" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Objective: determine if tissue-specific promoter marks (e.g.Â H3K27ac)
are often near genes that are expressed in a tissue-specific manner.</p>
<p>We will load expression data from the GTEx project <span class="citation">(<a href="#ref-gtex" role="doc-biblioref">Lonsdale and others 2013</a>)</span>, which gives
median expression in TPM for each tissue. We will use H3K27ac ChIP-seq
data from the ENCODE project <span class="citation">(<a href="#ref-encode" role="doc-biblioref">Dunham and others 2012</a>)</span>.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="promoter-marks.html#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb57-2"><a href="promoter-marks.html#cb57-2" aria-hidden="true" tabindex="-1"></a>file <span class="ot">&lt;-</span> <span class="st">&quot;data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz&quot;</span></span>
<span id="cb57-3"><a href="promoter-marks.html#cb57-3" aria-hidden="true" tabindex="-1"></a>gtex <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(file, <span class="at">skip=</span><span class="dv">2</span>)</span></code></pre></div>
<p>We select two tissues, bladder and kidney, and convert the data from a
wide format into a tidy format.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="promoter-marks.html#cb58-1" aria-hidden="true" tabindex="-1"></a>tissues <span class="ot">&lt;-</span> gtex <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(Name, Bladder, Kidney...Cortex) <span class="sc">%&gt;%</span></span>
<span id="cb58-2"><a href="promoter-marks.html#cb58-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">gene =</span> Name, <span class="at">Kidney =</span> Kidney...Cortex) <span class="sc">%&gt;%</span></span>
<span id="cb58-3"><a href="promoter-marks.html#cb58-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">sub</span>(<span class="st">&quot;</span><span class="sc">\\</span><span class="st">..*&quot;</span>,<span class="st">&quot;&quot;</span>,gene)) <span class="sc">%&gt;%</span></span>
<span id="cb58-4"><a href="promoter-marks.html#cb58-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">!</span>gene, <span class="at">names_to=</span><span class="st">&quot;tissue&quot;</span>, <span class="at">values_to=</span><span class="st">&quot;tpm&quot;</span>)</span></code></pre></div>
<p>Now define two vectors of genes that are specific to bladder and kidney:</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="promoter-marks.html#cb59-1" aria-hidden="true" tabindex="-1"></a>bladder_expr <span class="ot">&lt;-</span> tissues <span class="sc">%&gt;%</span></span>
<span id="cb59-2"><a href="promoter-marks.html#cb59-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(tissue <span class="sc">==</span> <span class="st">&quot;Bladder&quot;</span> <span class="sc">&amp;</span> tpm <span class="sc">&gt;</span> <span class="dv">10</span>) <span class="sc">%&gt;%</span></span>
<span id="cb59-3"><a href="promoter-marks.html#cb59-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">pull</span>(gene)</span>
<span id="cb59-4"><a href="promoter-marks.html#cb59-4" aria-hidden="true" tabindex="-1"></a>kidney_expr <span class="ot">&lt;-</span> tissues <span class="sc">%&gt;%</span></span>
<span id="cb59-5"><a href="promoter-marks.html#cb59-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(tissue <span class="sc">==</span> <span class="st">&quot;Kidney&quot;</span> <span class="sc">&amp;</span> tpm <span class="sc">&gt;</span> <span class="dv">10</span>) <span class="sc">%&gt;%</span></span>
<span id="cb59-6"><a href="promoter-marks.html#cb59-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">pull</span>(gene)</span>
<span id="cb59-7"><a href="promoter-marks.html#cb59-7" aria-hidden="true" tabindex="-1"></a>int <span class="ot">&lt;-</span> <span class="fu">intersect</span>(bladder_expr, kidney_expr)</span>
<span id="cb59-8"><a href="promoter-marks.html#cb59-8" aria-hidden="true" tabindex="-1"></a>bladder_expr <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(bladder_expr, int)</span>
<span id="cb59-9"><a href="promoter-marks.html#cb59-9" aria-hidden="true" tabindex="-1"></a>kidney_expr <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(kidney_expr, int)</span>
<span id="cb59-10"><a href="promoter-marks.html#cb59-10" aria-hidden="true" tabindex="-1"></a><span class="co"># save(bladder_expr, kidney_expr, file=&quot;data/bladder_kidney_expr.rda&quot;)</span></span></code></pre></div>
<p>Next, use an existing TxDb to locate these genes in the genomes. While
we usually recommend to use GENCODE genes for human analysis, because
the ENCODE chromatin modification peak files on AnnotationHub are in
hg19, we use the UCSC hg19 genes here for simplicity of the code:</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="promoter-marks.html#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)</span>
<span id="cb60-2"><a href="promoter-marks.html#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span></code></pre></div>
<p>Add the ENSEMBL ID and pull out the two tissue-specific sets.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="promoter-marks.html#cb61-1" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">genes</span>(TxDb.Hsapiens.UCSC.hg19.knownGene)</span></code></pre></div>
<pre><code>##   403 genes were dropped because they have exons located on both strands of the same
##   reference sequence or on more than one reference sequence, so cannot be represented by a
##   single genomic range.
##   Use &#39;single.strand.genes.only=FALSE&#39; to get all the genes in a GRangesList object, or use
##   suppressMessages() to suppress this message.</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="promoter-marks.html#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyranges)</span>
<span id="cb63-2"><a href="promoter-marks.html#cb63-2" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> g <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">ensembl =</span> <span class="fu">mapIds</span>(org.Hs.eg.db, gene_id, <span class="st">&quot;ENSEMBL&quot;</span>, <span class="st">&quot;ENTREZID&quot;</span>))</span>
<span id="cb63-3"><a href="promoter-marks.html#cb63-3" aria-hidden="true" tabindex="-1"></a>bladder_g <span class="ot">&lt;-</span> g <span class="sc">%&gt;%</span> <span class="fu">filter</span>(ensembl <span class="sc">%in%</span> bladder_expr)</span>
<span id="cb63-4"><a href="promoter-marks.html#cb63-4" aria-hidden="true" tabindex="-1"></a>kidney_g <span class="ot">&lt;-</span> g <span class="sc">%&gt;%</span> <span class="fu">filter</span>(ensembl <span class="sc">%in%</span> kidney_expr)</span></code></pre></div>
<p>Finally we combine the two sets with <code>bind_ranges</code>, and we change
the feature size from the whole gene extent (the range from the
leftmost exon to rightmost exon), to just the TSS, using <code>anchor_5p</code>
and <code>mutate</code>.</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="promoter-marks.html#cb64-1" aria-hidden="true" tabindex="-1"></a>tss <span class="ot">&lt;-</span> <span class="fu">bind_ranges</span>(<span class="at">bladder=</span>bladder_g,</span>
<span id="cb64-2"><a href="promoter-marks.html#cb64-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">kidney=</span>kidney_g,</span>
<span id="cb64-3"><a href="promoter-marks.html#cb64-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.id=</span><span class="st">&quot;gtissue&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb64-4"><a href="promoter-marks.html#cb64-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">anchor_5p</span>() <span class="sc">%&gt;%</span></span>
<span id="cb64-5"><a href="promoter-marks.html#cb64-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">width=</span><span class="dv">1</span>)</span></code></pre></div>
<p>Now we will obtain the H3K27ac peak sets:</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="promoter-marks.html#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(AnnotationHub)</span>
<span id="cb65-2"><a href="promoter-marks.html#cb65-2" aria-hidden="true" tabindex="-1"></a>ah <span class="ot">&lt;-</span> <span class="fu">AnnotationHub</span>()</span>
<span id="cb65-3"><a href="promoter-marks.html#cb65-3" aria-hidden="true" tabindex="-1"></a><span class="co"># query(ah, c(&quot;Homo sapiens&quot;, &quot;bladder&quot;, &quot;H3K27ac&quot;, &quot;narrowPeak&quot;))</span></span>
<span id="cb65-4"><a href="promoter-marks.html#cb65-4" aria-hidden="true" tabindex="-1"></a>bladder_pks <span class="ot">&lt;-</span> ah[[<span class="st">&quot;AH44180&quot;</span>]]</span>
<span id="cb65-5"><a href="promoter-marks.html#cb65-5" aria-hidden="true" tabindex="-1"></a><span class="co"># query(ah, c(&quot;Homo sapiens&quot;, &quot;kidney&quot;, &quot;H3K27ac&quot;, &quot;narrowPeak&quot;))</span></span>
<span id="cb65-6"><a href="promoter-marks.html#cb65-6" aria-hidden="true" tabindex="-1"></a>kidney_pks <span class="ot">&lt;-</span> ah[[<span class="st">&quot;AH43443&quot;</span>]]</span>
<span id="cb65-7"><a href="promoter-marks.html#cb65-7" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(bladder_pks, kidney_pks, <span class="at">file=</span><span class="st">&quot;data/peaks.rda&quot;</span>)</span></code></pre></div>
<p>We download these and scale so they have the same 90% quantile of <code>signalValue</code>.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="promoter-marks.html#cb66-1" aria-hidden="true" tabindex="-1"></a>ninety <span class="ot">&lt;-</span> <span class="cf">function</span>(x) <span class="fu">quantile</span>(x, .<span class="dv">9</span>, <span class="at">names=</span><span class="cn">FALSE</span>)</span>
<span id="cb66-2"><a href="promoter-marks.html#cb66-2" aria-hidden="true" tabindex="-1"></a>bladder_pks <span class="ot">&lt;-</span> bladder_pks <span class="sc">%&gt;%</span></span>
<span id="cb66-3"><a href="promoter-marks.html#cb66-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">signal =</span> signalValue <span class="sc">/</span> <span class="fu">ninety</span>(signalValue))</span>
<span id="cb66-4"><a href="promoter-marks.html#cb66-4" aria-hidden="true" tabindex="-1"></a>kidney_pks <span class="ot">&lt;-</span> kidney_pks <span class="sc">%&gt;%</span></span>
<span id="cb66-5"><a href="promoter-marks.html#cb66-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">signal =</span> signalValue <span class="sc">/</span> <span class="fu">ninety</span>(signalValue))</span></code></pre></div>
<p>Combine the peaks from bladder and kidney, filter to those with &lt; 0.1%
FDR, and center the peak on the summit (the <code>peak</code> column gives the
shift from the left side to the summit).</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="promoter-marks.html#cb67-1" aria-hidden="true" tabindex="-1"></a>pks <span class="ot">&lt;-</span> <span class="fu">bind_ranges</span>(<span class="at">bladder=</span>bladder_pks,</span>
<span id="cb67-2"><a href="promoter-marks.html#cb67-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">kidney=</span>kidney_pks,</span>
<span id="cb67-3"><a href="promoter-marks.html#cb67-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">.id=</span><span class="st">&quot;ptissue&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb67-4"><a href="promoter-marks.html#cb67-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(qValue <span class="sc">&gt;</span> <span class="dv">3</span>, width <span class="sc">&lt;=</span> <span class="dv">1000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb67-5"><a href="promoter-marks.html#cb67-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">start =</span> start <span class="sc">+</span> peak) <span class="sc">%&gt;%</span></span>
<span id="cb67-6"><a href="promoter-marks.html#cb67-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>peak) <span class="sc">%&gt;%</span></span>
<span id="cb67-7"><a href="promoter-marks.html#cb67-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">width =</span> <span class="dv">1</span>)</span></code></pre></div>
<p>Finally, once we have two tidy range sets, we can perform the analysis
by a join, followed by two lines that take care of multiple overlaps,
followed by two lines that give us our tallies of interest.</p>
<p>It appears that tissue-specific peaks are enriched near the
tissue-specific genes for both bladder and kidney.</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="promoter-marks.html#cb68-1" aria-hidden="true" tabindex="-1"></a>tss <span class="sc">%&gt;%</span></span>
<span id="cb68-2"><a href="promoter-marks.html#cb68-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">join_overlap_left</span>(pks, <span class="at">maxgap=</span><span class="dv">500</span>) <span class="sc">%&gt;%</span></span>
<span id="cb68-3"><a href="promoter-marks.html#cb68-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(ptissue) <span class="sc">%&gt;%</span> <span class="co"># within peak tissue...</span></span>
<span id="cb68-4"><a href="promoter-marks.html#cb68-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">duplicated</span>(gene_id)) <span class="sc">%&gt;%</span> <span class="co"># ...just take the first overlap per gene</span></span>
<span id="cb68-5"><a href="promoter-marks.html#cb68-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gtissue, ptissue) <span class="sc">%&gt;%</span></span>
<span id="cb68-6"><a href="promoter-marks.html#cb68-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">count =</span> <span class="fu">n</span>())</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 3 columns
##   gtissue     ptissue     count
##     &lt;Rle&gt; &lt;character&gt; &lt;integer&gt;
## 1 bladder     bladder      2163
## 2 bladder      kidney      1568
## 3 bladder          NA       859
## 4  kidney     bladder       103
## 5  kidney      kidney       208
## 6  kidney          NA       242</code></pre>
<p>The above number could also be found with four <code>countOverlaps</code> calls,
by considering all four pairs of overlaps of the two sets of genes and
peaks.</p>
<p>Another way to avoid counting overlaps more than once per gene is to
use the <em>plyranges</em> function, <code>n_distinct()</code>:</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="promoter-marks.html#cb70-1" aria-hidden="true" tabindex="-1"></a>tss <span class="sc">%&gt;%</span></span>
<span id="cb70-2"><a href="promoter-marks.html#cb70-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">join_overlap_left</span>(pks, <span class="at">maxgap=</span><span class="dv">500</span>) <span class="sc">%&gt;%</span></span>
<span id="cb70-3"><a href="promoter-marks.html#cb70-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gtissue, ptissue) <span class="sc">%&gt;%</span></span>
<span id="cb70-4"><a href="promoter-marks.html#cb70-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">count =</span> <span class="fu">n_distinct</span>(gene_id))</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 3 columns
##   gtissue     ptissue     count
##     &lt;Rle&gt; &lt;character&gt; &lt;integer&gt;
## 1 bladder     bladder      2163
## 2 bladder      kidney      1568
## 3 bladder          NA       859
## 4  kidney     bladder       103
## 5  kidney      kidney       208
## 6  kidney          NA       242</code></pre>
<p>If we want more information per gene, e.g.Â suppose we want to compute
the average signal per gene of peaks nearby, we need to group twice,
once also by gene ID, and the second time integrating over gene
ID. While here we add a few more lines of code, performing such an
operation with base Bioconductor functions would require adding code
to perform the loops, adding many intermediate variables to store
results, etc.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="promoter-marks.html#cb72-1" aria-hidden="true" tabindex="-1"></a>tss <span class="sc">%&gt;%</span></span>
<span id="cb72-2"><a href="promoter-marks.html#cb72-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">join_overlap_left</span>(pks, <span class="at">maxgap=</span><span class="dv">500</span>) <span class="sc">%&gt;%</span></span>
<span id="cb72-3"><a href="promoter-marks.html#cb72-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gtissue, ptissue, gene_id) <span class="sc">%&gt;%</span> <span class="co"># need per gene stats</span></span>
<span id="cb72-4"><a href="promoter-marks.html#cb72-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">num_overlaps =</span> <span class="fu">n</span>(), <span class="at">signal =</span> <span class="fu">mean</span>(signal)) <span class="sc">%&gt;%</span></span>
<span id="cb72-5"><a href="promoter-marks.html#cb72-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> <span class="co"># DataFrame to tibble for further processing</span></span>
<span id="cb72-6"><a href="promoter-marks.html#cb72-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gtissue, ptissue) <span class="sc">%&gt;%</span></span>
<span id="cb72-7"><a href="promoter-marks.html#cb72-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">sum_any_overlaps =</span> <span class="fu">sum</span>(num_overlaps <span class="sc">&gt;</span> <span class="dv">0</span>),</span>
<span id="cb72-8"><a href="promoter-marks.html#cb72-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_signal=</span><span class="fu">mean</span>(signal))</span></code></pre></div>
<pre><code>## `summarise()` has grouped output by &#39;gtissue&#39;. You can override using the `.groups` argument.</code></pre>
<pre><code>## # A tibble: 6 Ã 4
## # Groups:   gtissue [2]
##   gtissue ptissue sum_any_overlaps mean_signal
##   &lt;chr&gt;   &lt;chr&gt;              &lt;int&gt;       &lt;dbl&gt;
## 1 bladder bladder             2163       1.18 
## 2 bladder kidney              1568       0.929
## 3 bladder &lt;NA&gt;                 859      NA    
## 4 kidney  bladder              103       0.991
## 5 kidney  kidney               208       0.714
## 6 kidney  &lt;NA&gt;                 242      NA</code></pre>
<p>Whatâs wrong with this analysis?</p>
<ul>
<li>We didnât figure out the expressed promoter, we just looked at the
left or rightmost isoform (for + or - strand genes, respectively).</li>
</ul>

</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-encode" class="csl-entry">
Dunham, I., and others. 2012. <span>â<span class="nocase">An integrated encyclopedia of DNA elements in the human genome</span>.â</span> <em>Nature</em> 489 (7414): 57â74.
</div>
<div id="ref-gtex" class="csl-entry">
Lonsdale, J., and others. 2013. <span>â<span class="nocase">The Genotype-Tissue Expression (GTEx) project</span>.â</span> <em>Nature Genetics</em> 45 (6): 580â85.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="compute-overlaps.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="bootstrap-overlap.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
