<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 SNP position in peaks | Tidy Ranges Tutorial</title>
  <meta name="description" content="<p>Basic examples of computing operations on genomic ranges using the
tidy data philosophy.</p>" />
  <meta name="generator" content="bookdown 0.32 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 SNP position in peaks | Tidy Ranges Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="<p>Basic examples of computing operations on genomic ranges using the
tidy data philosophy.</p>" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 SNP position in peaks | Tidy Ranges Tutorial" />
  
  <meta name="twitter:description" content="<p>Basic examples of computing operations on genomic ranges using the
tidy data philosophy.</p>" />
  

<meta name="author" content="nullranges devel team" />


<meta name="date" content="2023-03-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="bootstrap-overlap.html"/>
<link rel="next" href="tidy-gene-plots.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Tidy Ranges Tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a></li>
<li class="chapter" data-level="1" data-path="join-is-an-overlap.html"><a href="join-is-an-overlap.html"><i class="fa fa-check"></i><b>1</b> Join is an overlap</a></li>
<li class="chapter" data-level="2" data-path="overlap-granges-with-plyranges.html"><a href="overlap-granges-with-plyranges.html"><i class="fa fa-check"></i><b>2</b> Overlap GRanges with plyranges</a></li>
<li class="chapter" data-level="3" data-path="tissue-specific-promoter-marks.html"><a href="tissue-specific-promoter-marks.html"><i class="fa fa-check"></i><b>3</b> Tissue-specific promoter marks</a></li>
<li class="chapter" data-level="4" data-path="bootstrap-overlap.html"><a href="bootstrap-overlap.html"><i class="fa fa-check"></i><b>4</b> Bootstrap overlap</a></li>
<li class="chapter" data-level="5" data-path="snp-position-in-peaks.html"><a href="snp-position-in-peaks.html"><i class="fa fa-check"></i><b>5</b> SNP position in peaks</a></li>
<li class="chapter" data-level="6" data-path="tidy-gene-plots.html"><a href="tidy-gene-plots.html"><i class="fa fa-check"></i><b>6</b> Tidy gene plots</a></li>
<li class="chapter" data-level="7" data-path="tidy-rna-seq-eda.html"><a href="tidy-rna-seq-eda.html"><i class="fa fa-check"></i><b>7</b> Tidy RNA-seq EDA</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tidy Ranges Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="snp-position-in-peaks" class="section level1 hasAnchor" number="5">
<h1><span class="header-section-number">Chapter 5</span> SNP position in peaks<a href="snp-position-in-peaks.html#snp-position-in-peaks" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Objective: determine the position of a set of SNPs in peaks, that is, determine the relative positions of those SNPs that overlap the peaks.</p>
<p>We again will use the ENCODE kidney H3K27ac ChIP-seq peaks used in the previous analysis <span class="citation">(Dunham and others <a href="#ref-encode" role="doc-biblioref">2012</a>)</span>. We will create some artificial SNPs: this analysis could generalize to any time that we have two ranges, where we are interested in the relative position of one set of ranges (SNPs) within the other set of ranges (peaks).</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="snp-position-in-peaks.html#cb110-1" aria-hidden="true"></a><span class="kw">library</span>(AnnotationHub)</span>
<span id="cb110-2"><a href="snp-position-in-peaks.html#cb110-2" aria-hidden="true"></a>ah &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()</span>
<span id="cb110-3"><a href="snp-position-in-peaks.html#cb110-3" aria-hidden="true"></a>kidney_pks &lt;-<span class="st"> </span>ah[[<span class="st">&quot;AH43443&quot;</span>]]</span></code></pre></div>
<p>We will filter the peaks to standard chromosomes, and include the same cutoffs we used in the previous analysis.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="snp-position-in-peaks.html#cb111-1" aria-hidden="true"></a><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(GenomeInfoDb))</span>
<span id="cb111-2"><a href="snp-position-in-peaks.html#cb111-2" aria-hidden="true"></a>pks &lt;-<span class="st"> </span>kidney_pks</span>
<span id="cb111-3"><a href="snp-position-in-peaks.html#cb111-3" aria-hidden="true"></a>pks &lt;-<span class="st"> </span><span class="kw">keepStandardChromosomes</span>(pks)</span></code></pre></div>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="snp-position-in-peaks.html#cb113-1" aria-hidden="true"></a><span class="kw">suppressPackageStartupMessages</span>({</span>
<span id="cb113-2"><a href="snp-position-in-peaks.html#cb113-2" aria-hidden="true"></a>  <span class="kw">library</span>(dplyr)</span>
<span id="cb113-3"><a href="snp-position-in-peaks.html#cb113-3" aria-hidden="true"></a>  <span class="kw">library</span>(tibble)</span>
<span id="cb113-4"><a href="snp-position-in-peaks.html#cb113-4" aria-hidden="true"></a>  <span class="kw">library</span>(plyranges)</span>
<span id="cb113-5"><a href="snp-position-in-peaks.html#cb113-5" aria-hidden="true"></a>})</span>
<span id="cb113-6"><a href="snp-position-in-peaks.html#cb113-6" aria-hidden="true"></a>q_thr &lt;-<span class="st"> </span><span class="dv">3</span></span>
<span id="cb113-7"><a href="snp-position-in-peaks.html#cb113-7" aria-hidden="true"></a>s_thr &lt;-<span class="st"> </span><span class="dv">9</span></span>
<span id="cb113-8"><a href="snp-position-in-peaks.html#cb113-8" aria-hidden="true"></a>pks &lt;-<span class="st"> </span>pks <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb113-9"><a href="snp-position-in-peaks.html#cb113-9" aria-hidden="true"></a><span class="st">  </span><span class="kw">filter</span>(qValue <span class="op">&gt;</span><span class="st"> </span>q_thr <span class="op">&amp;</span><span class="st"> </span>signalValue <span class="op">&gt;</span><span class="st"> </span>s_thr) <span class="op">%&gt;%</span></span>
<span id="cb113-10"><a href="snp-position-in-peaks.html#cb113-10" aria-hidden="true"></a><span class="st">  </span><span class="kw">sort</span>()</span></code></pre></div>
<p>A histogram of peak width:</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="snp-position-in-peaks.html#cb114-1" aria-hidden="true"></a><span class="kw">library</span>(ggplot2)</span>
<span id="cb114-2"><a href="snp-position-in-peaks.html#cb114-2" aria-hidden="true"></a>pks <span class="op">%&gt;%</span><span class="st"> </span>as_tibble <span class="op">%&gt;%</span></span>
<span id="cb114-3"><a href="snp-position-in-peaks.html#cb114-3" aria-hidden="true"></a><span class="st">  </span><span class="kw">filter</span>(width <span class="op">&lt;</span><span class="st"> </span><span class="dv">5000</span>) <span class="op">%&gt;%</span></span>
<span id="cb114-4"><a href="snp-position-in-peaks.html#cb114-4" aria-hidden="true"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(width)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb114-5"><a href="snp-position-in-peaks.html#cb114-5" aria-hidden="true"></a><span class="st">  </span><span class="kw">geom_histogram</span>()</span></code></pre></div>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="snp-position-in-peaks_files/figure-html/peak-widths-1.png" width="672" /></p>
<p>Let’s subset to peaks that are less than 2000 in width:</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="snp-position-in-peaks.html#cb116-1" aria-hidden="true"></a>pks &lt;-<span class="st"> </span>pks <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(width <span class="op">&lt;</span><span class="st"> </span><span class="dv">2000</span>)</span></code></pre></div>
<p>Now, we may be interested in SNPs that are associated with regulatory function. While we could find these from a database of chromatin QTL or functionally validated variants, here we will just create a simulated set of SNPs for demonstration.</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="snp-position-in-peaks.html#cb117-1" aria-hidden="true"></a><span class="kw">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb117-2"><a href="snp-position-in-peaks.html#cb117-2" aria-hidden="true"></a>snps &lt;-<span class="st"> </span>pks <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb117-3"><a href="snp-position-in-peaks.html#cb117-3" aria-hidden="true"></a><span class="st">  </span><span class="kw">slice</span>(<span class="kw">sample.int</span>(<span class="kw">n</span>(), <span class="dv">2000</span>)) <span class="op">%&gt;%</span></span>
<span id="cb117-4"><a href="snp-position-in-peaks.html#cb117-4" aria-hidden="true"></a><span class="st">  </span><span class="kw">anchor_5p</span>() <span class="op">%&gt;%</span></span>
<span id="cb117-5"><a href="snp-position-in-peaks.html#cb117-5" aria-hidden="true"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">start=</span>start <span class="op">+</span><span class="st"> </span><span class="kw">floor</span>(<span class="kw">runif</span>(<span class="kw">n</span>(),<span class="dv">0</span>,width))) <span class="op">%&gt;%</span></span>
<span id="cb117-6"><a href="snp-position-in-peaks.html#cb117-6" aria-hidden="true"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">width=</span><span class="dv">1</span>)</span>
<span id="cb117-7"><a href="snp-position-in-peaks.html#cb117-7" aria-hidden="true"></a>snps</span></code></pre></div>
<pre><code>## GRanges object with 2000 ranges and 6 metadata columns:
##          seqnames    ranges strand |        name     score signalValue
##             &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;   &lt;numeric&gt;
##      [1]     chr1 222886667      * |    Rank_786       943    28.06788
##      [2]    chr17   1251302      * |  Rank_11154       238    10.45065
##      [3]     chr9  75179753      * |   Rank_8060       305    12.52016
##      [4]     chrX  70401571      * |   Rank_9602       266    11.38197
##      [5]    chr20  37615472      * |  Rank_14922       190     9.52741
##      ...      ...       ...    ... .         ...       ...         ...
##   [1996]     chr9  14316343      * |  Rank_15136       188     9.11715
##   [1997]    chr18  74800298      * |  Rank_15685       181     9.02626
##   [1998]    chr19  52641902      * |   Rank_2464       606    19.64891
##   [1999]     chr2 128642862      * |  Rank_15514       184     9.18719
##   [2000]     chr6 151005236      * |  Rank_13672       203     9.07738
##             pValue    qValue      peak
##          &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
##      [1]   94.3437   89.9679       267
##      [2]   23.8843   21.0424       458
##      [3]   30.5751   27.5376       616
##      [4]   26.6991   23.7679       455
##      [5]   19.0203   16.3297       414
##      ...       ...       ...       ...
##   [1996]   18.8452   16.1647       461
##   [1997]   18.1808   15.5192       173
##   [1998]   60.6790   56.9445      1281
##   [1999]   18.4512   15.7837       125
##   [2000]   20.3511   17.6170       887
##   -------
##   seqinfo: 25 sequences (1 circular) from hg19 genome</code></pre>
<p>These should all overlap peaks:</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="snp-position-in-peaks.html#cb119-1" aria-hidden="true"></a>snps <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb119-2"><a href="snp-position-in-peaks.html#cb119-2" aria-hidden="true"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">n_overlaps =</span> <span class="kw">count_overlaps</span>(., pks)) <span class="op">%&gt;%</span></span>
<span id="cb119-3"><a href="snp-position-in-peaks.html#cb119-3" aria-hidden="true"></a><span class="st">  </span><span class="kw">summarize</span>(<span class="dt">tab=</span><span class="kw">table</span>(n_overlaps))</span></code></pre></div>
<pre><code>## DataFrame with 1 row and 2 columns
##   tab.n_overlaps  tab.Freq
##         &lt;factor&gt; &lt;integer&gt;
## 1              1      2000</code></pre>
<p>We now have a bit of a hack for this analysis: we add the start position and width of the peak as addition columns of metadata. This is because otherwise, we will lose the start position when we perform the overlap of SNPs in peaks, and we need it for later.</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="snp-position-in-peaks.html#cb121-1" aria-hidden="true"></a>pks_trim &lt;-<span class="st"> </span>pks <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb121-2"><a href="snp-position-in-peaks.html#cb121-2" aria-hidden="true"></a><span class="st">  </span><span class="kw">select</span>(name) <span class="op">%&gt;%</span></span>
<span id="cb121-3"><a href="snp-position-in-peaks.html#cb121-3" aria-hidden="true"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">peak_start=</span>start, <span class="dt">peak_width=</span>width)</span></code></pre></div>
<p>We can then obtain overlaps of SNPs with peaks, and add the relative position of the SNP within the peak extent. We use 1-based indexing, such that if the SNP is the same as the leftmost basepair of the peak, it gets counted as position 1.</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="snp-position-in-peaks.html#cb122-1" aria-hidden="true"></a>o &lt;-<span class="st"> </span>snps <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb122-2"><a href="snp-position-in-peaks.html#cb122-2" aria-hidden="true"></a><span class="st">  </span><span class="kw">join_overlap_inner</span>(pks_trim) <span class="op">%&gt;%</span></span>
<span id="cb122-3"><a href="snp-position-in-peaks.html#cb122-3" aria-hidden="true"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">rel_pos =</span> start <span class="op">-</span><span class="st"> </span>peak_start <span class="op">+</span><span class="st"> </span><span class="dv">1</span>,</span>
<span id="cb122-4"><a href="snp-position-in-peaks.html#cb122-4" aria-hidden="true"></a>         <span class="dt">rel_frac =</span> (rel_pos <span class="op">-</span><span class="st"> </span><span class="dv">1</span>) <span class="op">/</span><span class="st"> </span>(peak_width <span class="op">-</span><span class="st"> </span><span class="dv">1</span> ))</span></code></pre></div>
<p>Check our expectations about these new columns:</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="snp-position-in-peaks.html#cb123-1" aria-hidden="true"></a><span class="kw">all</span>(o<span class="op">$</span>rel_pos <span class="op">&gt;=</span><span class="st"> </span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="snp-position-in-peaks.html#cb125-1" aria-hidden="true"></a><span class="kw">all</span>(o<span class="op">$</span>rel_frac <span class="op">&gt;=</span><span class="st"> </span><span class="dv">0</span>)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="snp-position-in-peaks.html#cb127-1" aria-hidden="true"></a><span class="kw">all</span>(o<span class="op">$</span>rel_frac <span class="op">&lt;=</span><span class="st"> </span><span class="dv">1</span>)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Finally, we can compute a histogram of where the SNPs fall in the peaks:</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="snp-position-in-peaks.html#cb129-1" aria-hidden="true"></a>o <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb129-2"><a href="snp-position-in-peaks.html#cb129-2" aria-hidden="true"></a><span class="st">  </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></span>
<span id="cb129-3"><a href="snp-position-in-peaks.html#cb129-3" aria-hidden="true"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(rel_frac)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb129-4"><a href="snp-position-in-peaks.html#cb129-4" aria-hidden="true"></a><span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">breaks=</span><span class="dv">0</span><span class="op">:</span><span class="dv">10</span><span class="op">/</span><span class="dv">10</span>)</span></code></pre></div>
<p><img src="snp-position-in-peaks_files/figure-html/snp-pos-hist-1.png" width="672" /></p>
<p>Now stratifying by width of peak:</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="snp-position-in-peaks.html#cb130-1" aria-hidden="true"></a><span class="kw">quantile</span>(o<span class="op">$</span>peak_width, <span class="dv">0</span><span class="op">:</span><span class="dv">3</span><span class="op">/</span><span class="dv">3</span>)</span></code></pre></div>
<pre><code>##        0% 33.33333% 66.66667%      100% 
##  246.0000  813.3333 1289.6667 1998.0000</code></pre>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="snp-position-in-peaks.html#cb132-1" aria-hidden="true"></a>o <span class="op">%&gt;%</span></span>
<span id="cb132-2"><a href="snp-position-in-peaks.html#cb132-2" aria-hidden="true"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">width_bin =</span> <span class="kw">cut</span>(peak_width,</span>
<span id="cb132-3"><a href="snp-position-in-peaks.html#cb132-3" aria-hidden="true"></a>                         <span class="dt">breaks=</span><span class="kw">c</span>(<span class="dv">200</span>,<span class="dv">800</span>,<span class="dv">1300</span>,<span class="dv">2000</span>),</span>
<span id="cb132-4"><a href="snp-position-in-peaks.html#cb132-4" aria-hidden="true"></a>                         <span class="dt">include.lowest=</span><span class="ot">TRUE</span>)) <span class="op">%&gt;%</span></span>
<span id="cb132-5"><a href="snp-position-in-peaks.html#cb132-5" aria-hidden="true"></a><span class="st">  </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></span>
<span id="cb132-6"><a href="snp-position-in-peaks.html#cb132-6" aria-hidden="true"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(rel_frac)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb132-7"><a href="snp-position-in-peaks.html#cb132-7" aria-hidden="true"></a><span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">breaks=</span><span class="dv">0</span><span class="op">:</span><span class="dv">10</span><span class="op">/</span><span class="dv">10</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb132-8"><a href="snp-position-in-peaks.html#cb132-8" aria-hidden="true"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>width_bin)</span></code></pre></div>
<p><img src="snp-position-in-peaks_files/figure-html/snp-pos-hist-facet-1.png" width="672" /></p>

</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references hanging-indent">
<div id="ref-encode">
<p>Dunham, I., and others. 2012. “An integrated encyclopedia of DNA elements in the human genome.” <em>Nature</em> 489 (7414): 57–74.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="bootstrap-overlap.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="tidy-gene-plots.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["tidy-ranges-tutorial.pdf", "tidy-ranges-tutorial.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
